### ISSUE TITLE
How to install Grocy on FreeNas FreeBSD 11.3 Jail using Nginx

### ISSUE THREAD

**Issue by mikepopa** [grocy/grocy#715]
I couldn't find a step by step guide on how to install on a FreeNas jail, so here is what worked for me.

Installing in a FreeBSD 11.2 did not work because I couldn't get php-fpm to start.
I was only able to install on a FreeBSD 11.3 Jail using Nginx. I'm sure it also works with Apache, feel free to try it out.

### 1. Enable SSH and root login
This step is optional. You can use the GUI shell or anything else. I have used Putty for better copy/paste into the config files, so I had to SSH in.
- Enable SSH Service:
	`sysrc sshd_enable=yes`

- To allow root login via SSH: 
 	`ee /etc/ssh/sshd_config` 
Uncomment the following and set to YES.
	`PermitRootLogin YES`
 Hit Escape, `a)` to leave the editor and `a)` to save changes.
	
- Start the SSHD service:
	`service sshd start`
- Change the root password if you need to:
	`passwd root`
	`<new root password>`
	`<confirm root password>`
- SSH into the Jail

### 2.  Update packages
`ee /etc/pkg/FreeBSD.conf`
- Change URL from quarterly to latest.
	Change this	`url: "pkg+http://pkg.FreeBSD.org/${ABI}/quarterly"`
	To this: `url: "pkg+http://pkg.FreeBSD.org/${ABI}/latest"`
	
	Hit Escape, `a)` to leave the editor and `a)` to save changes.
	
- Update packages and answer `Y` if asked to install the package manager.
	`pkg update`
	

### 3. Install utilities, Nginx, PHP and PHP extensions*:
`pkg install -y nano unzip wget nginx php73 php73-sqlite3 php73-filter php73-pdo_sqlite php73-json php73-tokenizer php73-iconv`

*I have found that these PHP extensions are needed by trial and error. This worked for me, but there might be more needed. If you get an error, just google it.

### 4. Enable auto-start for Nginx and PHP-FPM

`sysrc nginx_enable=yes`
`service nginx start`
`sysrc php_fpm_enable=yes`
`service php-fpm start`

### 5. Verify that the server works.
Navigate to your IP/domain in a browser. You should see the Nginx welcome page

### 6. Configure  PHP

`nano /usr/local/etc/php-fpm.d/www.conf`

- Replace the TCP socket with UNIX socket.
Change this:
`listen = 127.0.0.1:9000`
To this:
`listen = /var/run/php-fpm.sock;`

- Uncomment the lines:
`listen.owner = www`
`listen.group = www`
`listen.mode = 0660`
	

- Save and close the config file. `Ctl + X` to exit and `Y` to save changes
- Restart PHP-fpm service.
`service php-fpm restart`

### 7. Configure Nginx:

- Back-up original config file.
`mv /usr/local/etc/nginx/nginx.conf /usr/local/etc/nginx/nginx.conf_BK`
- You should change the IP to your IP/domain, but everything else should match.
`nano /usr/local/etc/nginx/nginx.conf`

```
user  www www;
worker_processes  1;

# This default error log path is compiled-in to make sure configuration parsing
# errors are logged somewhere, especially during unattended boot when stderr
# isn't normally logged anywhere. This path will be touched on every nginx
# start regardless of error log location configured here. See
# https://trac.nginx.org/nginx/ticket/147 for more info.
#
#error_log  /var/log/nginx/error.log;
#

#pid        logs/nginx.pid;


events {
    worker_connections  1024;
}


http {
    include       mime.types;
    default_type  application/octet-stream;

    #log_format  main  '$remote_addr - $remote_user [$time_local] "$request" '
    #                  '$status $body_bytes_sent "$http_referer" '
    #                  '"$http_user_agent" "$http_x_forwarded_for"';

    #access_log  logs/access.log  main;

    sendfile        on;
    #tcp_nopush     on;

    #keepalive_timeout  0;
    #keepalive_timeout  65;
    gzip  on;

server {
    server_name 192.168.1.133; # use domain name if you have here
    access_log  /var/log/nginx/192.168.1.133.access.log;
    error_log  /var/log/nginx/192.168.1.133.error.log;

    root /usr/local/www/grocy/public;
    index index.html index.htm index.php;
    # php config let nginx talk to php socket

        location / {
        try_files $uri /index.php;
        }
        location ~ [^/]\.php(/|$) {
        fastcgi_split_path_info ^(.+?\.php)(/.*)$;
        if (!-f $document_root$fastcgi_script_name) {
                return 404;
        }

        # Mitigate https://httpoxy.org/ vulnerabilities
         fastcgi_param HTTP_PROXY "";

        fastcgi_pass unix:/var/run/php-fpm.sock;
        fastcgi_index index.php;

        # include the fastcgi_param setting
        include fastcgi_params;

        # SCRIPT_FILENAME parameter is used for PHP FPM determining
        # the script name.
        fastcgi_param  SCRIPT_FILENAME   $document_root$fastcgi_script_name;
    }
}


    # another virtual host using mix of IP-, name-, and port-based configuration
    #
    #server {
    #    listen       8000;
    #    listen       somename:8080;
    #    server_name  somename  alias  another.alias;

    #    location / {
    #        root   html;
    #        index  index.html index.htm;
    #    }
    #}


    # HTTPS server
    #
    #server {
    #    listen       443 ssl;
    #    server_name  localhost;

    #    ssl_certificate      cert.pem;
    #    ssl_certificate_key  cert.key;

    #    ssl_session_cache    shared:SSL:1m;
    #    ssl_session_timeout  5m;

    #    ssl_ciphers  HIGH:!aNULL:!MD5;
    #    ssl_prefer_server_ciphers  on;

    #    location / {
    #        root   html;
    #        index  index.html index.htm;
    #    }
    #}

}
```

- Restart Nginx

`service nginx restart`

### 8. Install Grocy and set the folder permissions

- Go to your web root
`cd /usr/local/www`

- Download the grocy zip 
`wget --no-check-certificate https://github.com/grocy/grocy/releases/download/v2.6.2/grocy_2.6.2.zip`

- Unzip it in a directory called grocy. Make sure it contains all the files and not just another sub-directory.
`unzip grocy_2.6.2.zip -d grocy`

- Remove the archive from the www folder.
`rm grocy_2.6.2.zip`

- Change the directory owner an make it writeable.
`chown -R www:www /usr/local/www & chmod -R 755 /usr/local/www`

- Copy the default Grocy config to the data folder and name it config.php.
`cp /usr/local/www/grocy/config-dist.php /usr/local/www/grocy/data/config.php`

- Make any modifications to your config.php
`nano /usr/local/www/grocy/data/config.php`

- If you are using the same folder structure as in this example, the default config.php will work fine. My path and URL settings look like this:
	`Setting('BASE_PATH', '');`
	`Setting('BASE_URL', '/');`

**Comment by berrnd** [comment]
Thanks a lot! :)

Will add a link to this on https://grocy.info/links.

**Comment by andrewm659** [comment]
Ok what if I want to upgrade this?  Won't it lay down a new grocy.db file?

**Comment by berrnd** [comment]
Please use the [subreddit](https://www.reddit.com/r/grocy/) for general questions/help as [mentioned in README](https://github.com/grocy/grocy#questions--help--bug-reporting--feature-requests).

And no, the `data/grocy.db` is never shipped in the release ZIP, but populated/migrated and therefore of course preserved during upgrades.

**Comment by ProducerMatt** [comment]
Hi there! I worked around a couple problems, I figured I'd document it here just in case it helps anyone.

I diverged a little from the tutorial because I wanted the latest verion (2.7.1). But when I got the release zip it had carriage returns in it, which I've never encountered in other release zips. Not sure why.

Regardless, from digging around on Google, it looked like a Git clone would _always_ have the correct line endings, so I did: `pkg install git` (say yes to a million packages) and then `git clone https://github.com/grocy/groc` in /usr/local/www directory.

So I got around the line ending problem, but didn't have the yarn and composer dependencies installed.

First I got the needed packages on my system. Hopefully I didn't miss anything. Lots of these OP already mentioned but I found a few more were needed for building.

```pkg install npm php73 php73-composer php73-ctype php73-fileinfo php73-filter php73-gd php73-hash php73-iconv php73-intl php73-json php73-mbstring php73-openssl php73-pdo php73-pdo_sqlite php73-phar php73-simplexml php73-sqlite3 php73-tokenizer yarn```

and then `service restart php-fpm` to make sure the addons are loaded.

Doublecheck to be sure you're in the root of the repo and then run `composer.phar install` and `yarn install`.

Then, set up the config.php in `data`, **create the folder `data/viewcache`,** and `chown -R www:www` /  `chmod -R 755`  as mentioned.

That folder `data/viewcache` needs creating before the app works, I'm not sure why it isn't made automatically, but what do I know.

Hope this saves someone a little time. ☺️
